Support downloading bootstrap from CI in shell scripts

jyn514 opened the pull request:
- Download the beta compiler toolchain in bootstrap if it doesn't yet exist
    
    This is needed for when the shell scripts bypass python altogether and run the downloaded bootstrap directly

- Add a new `bootstrap-shim` program and distribute it on nightly

- Support downloading bootstrap from CI in the shell script entrypoints

  This completely bypasses the python scripts when bootstrap hasn't been modified.

I think we're going to need a way to opt-out of this, probably a new option in config.toml.

Helps with https://github.com/rust-lang/rust/issues/94829.

r? @Mark-Simulacrum

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3123990380/jobs/5066996263) [(plain)](https://github.com/rust-lang/rust/commit/366399aa40256e99ae96bb026e65eac4fef61b4a/checks/8540099795/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

 finished in 3.494 seconds
Build completed successfully in 0:35:03
+ ../x --stage 2 test src/test/mir-opt --host= --target=i686-unknown-linux-gnu
../x: 54: rustc: not found
downloading https://ci-artifacts.rust-lang.org/rustc-builds/4652f5eb25b76c24fc40b7267d87a07d5876901f/bootstrap-nightly-.tar.xz
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
```

</details>

jyn514 commented:
huh, that worked quicker than I expected

Mark-Simulacrum commented on src/bootstrap/config.rs:
```diff
@@ -910,10 +913,12 @@ impl Config {
+        config.initial_rustc = build.rustc.map(PathBuf::from).unwrap_or_else(|| {
+            let build = crate::Build::new(config.clone());
+            let builder = Builder::new(&build);
+            download_beta_toolchain(&builder);
```
This feels pretty suspicious. I don't think we should need to construct a full Build/Builder to download these binaries, what are they depending on that causes this? Can we move those functions onto Config?

jyn514 replied:
There's a *lot* of functions defined on Builder but not Config; no objection to moving them, but it's quite a large diff. We need to move:
- fix_bin_or_dylib
- download_component
- `verbose` (I'd keep a helper on Builder that calls Config::verbose for this)
- download_http_with_retries
- unpack
- verify
- tempdir
- `try_run` (I'd keep a helper)
- `check_run` (helper)
- `create` (helper)
- `remove` (helper)

and we might need some more after that, I don't know everything those functions call internally.

Hmm, I guess we need all that anyway to avoid pulling `Builder` into the shim; ok, seems worthwhile then.

Mark-Simulacrum commented on src/bootstrap/lib.rs:
```diff
@@ -426,7 +426,8 @@ impl Build {
         let in_tree_llvm_info = channel::GitInfo::new(false, &src.join("src/llvm-project"));
 
-        let initial_target_libdir_str = if config.dry_run {
+        let initial_target_libdir_str = if config.dry_run || config.initial_rustc == PathBuf::new()
```
I think this is caused by the Build::new construction I commented on before, which is why I was suspicious of it. :)

Mark-Simulacrum commented:
hey -- reviewing https://github.com/rust-lang/rust/pull/102282/ now, I'm feeling like maybe we really should have a shim that manages much of the complexity here.

I feel like my envisioning is that our shell scripts should be:

* check if rustup is installed, if so, run `rustup component add bootstrap-shim`, and execute that (rustup which rustc-bootstrap-shim, run that). That shim can be from any version of Rust (obviously to start just nightly but the shim should be pretty simple).
* the shim will take care of the git detection etc and can be compiled with a clap dependency so we can easily support --build-dir and all that stuff. I don't think there's a problem with us parsing config.toml, though it does lock down the compatibility of a subset of fields (we'll probably want to parse into dedicated shim structs that are a subset of the ones in bootstrap proper).

I think a reasonable alternative is that we expect a working Rust installation (not necessarily through rustup) and use that to compile a cargo-less src/bootstrap/entrypoint.rs that can be a shared executable and frankly is much nicer than shell or powershell (at least to my eyes), even without dependencies.

I am feeling like we ought to have a living document which we put our plan into and link from all of these PRs, because it seems to me that we are constantly re-debating these points and given my limited time I definitely am reinventing wheels since I can't keep all the context in my head.

pietroalbini commented:
> Commit authors other than `bors@rust-lang.org` are not supported. Fixing this is possible, but will require `jq` to be installed, which I don't want to require yet. I think doing feature detection (`command -v jq`) shouldn't be too bad, but if @pietroalbini doesn't plan to use the shell scripts, I don't think it really matters.

This would indeed be pretty annoying :(

A solution that doesn't involve `jq` would be to move the *configuration* out of `stage0.json` into a plaintext file with an easy enough syntax to parse with `bash`, something like:

```
dist_server https://static.rust-lang.org
artifacts_server https://ci-artifacts.rust-lang.org/rustc-builds
artifacts_with_llvm_assertions_server https://ci-artifacts.rust-lang.org/rustc-builds-alt
git_merge_commit_email bors@rust-lang.org
nightly_branch master
```

bors commented:
:umbrella: The latest upstream changes (presumably #100557) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 commented on x:
```diff
@@ -15,19 +21,65 @@ realpath() {
+fi
+
+latest_commit=$(git rev-list HEAD --author=bors@rust-lang.org --max-count 1)
+if git diff-index --quiet $latest_commit -- src/bootstrap; then
```
Should this be conditional on a config.toml config? A distro probably doesn't want to trigger this logic even if according to git they didn't change bootstrap.

jyn514 commented:
> I am feeling like we ought to have a living document which we put our plan into and link from all of these PRs, because it seems to me that we are constantly re-debating these points and given my limited time I definitely am reinventing wheels since I can't keep all the context in my head.

Sorry I keep changing the plan here. I've written up a short document with your `bootstrap-shim` suggestion: https://hackmd.io/@jynelson/S1Aa4cozo/edit. I'm going to implement those steps here.

If it helps, I think this should be the last PR where we ever have to rehash these details :) after this, the only steps left are to support downloadable bootstrap in python and update the documentation.

> I think a reasonable alternative is that we expect a working Rust installation (not necessarily through rustup) and use that to compile a cargo-less src/bootstrap/entrypoint.rs that can be a shared executable and frankly is much nicer than shell or powershell (at least to my eyes), even without dependencies.

Can you talk about this some more? This is the first time I've heard the suggestion - I think it would be somewhat limiting since it would mean we can't use `serde_toml` to parse the config file. I'm also unsure what running it would look like? I guess we could have a cargo alias, but people seemed reasonably happy with the shell scripts, and now that they're going to stay quite small, I'm not sure what benefit we'd get.

jyn514 commented:
I realized that sharing the code between the shim and the main binary without compiling it twice will be annoying; I've updated the hackmd (https://hackmd.io/j-XXBeYERuajJknd6zErPA) with a plan for doing that.

bjorn3 commented:
Did you see the comments I left on that hackmd?

jyn514 commented:
I didn't, thank you! Left a few responses :)

Mark-Simulacrum commented:
Left a comment on the HackMD -- but probably the plan seems broadly good.

jyn514 committed:
[wip] Add bootstrap-shim and package it on nightly

- Pass `dist bootstrap-shim` explicitly in CI

 This makes it possible to run `dist bootstrap` without also building
 bootstrap-shim, and more importantly works around a bug where currently
 two steps aren't allowed to have the same path.

- Add `check::BootstrapShim`
- [wip] start unifying parsing for Config and MinimalConfig

jyn514 commented on src/bootstrap/bin/bootstrap-shim.rs:
```diff
@@ -0,0 +1,27 @@
+use std::process::Command;
+
+use bootstrap::Config;
```
this is unfortunate; it depends on the entire bootstrap library, not just the parts it needs. I tried splitting them apart but it was taking much more time than I expected. Might make more sense to do it in a follow-up?

jyn514 replied:
fixing this would sidestep the problem in 
https://github.com/rust-lang/rust/pull/102282#discussion_r1024781628 though

jyn514 replied:
hmm, so here's one idea that could avoid depending on so much of bootstrap: add a new `mod min_config` that's still part of the main bootstrap library, and have bootstrap-shim only depend on things that are accessible through `bootstrap::min_config`, with a comment not to add more imports. Then we can go back and split min_config into a separate crate later, without it blocking the rest of the work.

jyn514 reviewed (commented):
this is still waiting on me, but it's a lot closer than before

jyn514 commented:
I am feeling less and less confident we can merge this change without something like https://github.com/rust-lang/rust/issues/102563 :/ for now I made the parsing for `min_config` and `config` entirely separate, so bugs in the former only affect the shim and not the main bootstrap binary, but it's not great that now there's duplicate code floating around ...

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3504966800/jobs/5870989644) [(plain)](https://github.com/rust-lang/rust/commit/25c6d7817df2b6f00a12692c18a992b65753ba02/checks/9594406062/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error: unused variable: `file`
   --> min_config.rs:237:13
    |
237 | fn get_toml(file: &Path) -> TomlConfig {
    |             ^^^^ help: if this is intentional, prefix it with an underscore: `_file`
    |
    = note: `-D unused-variables` implied by `-D warnings`

error[E0451]: field `minimal_config` of struct `config::Config` is private
   |
   |
21 |         ..Config::parse(&["check".to_owned()])
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ field `minimal_config` is private

error[E0451]: field `minimal_config` of struct `config::Config` is private
    |
    |
177 |         let config = Config { stage: 0, ..configure("build", &["A"], &["A"]) };
    |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ field `minimal_config` is private

error[E0451]: field `minimal_config` of struct `config::Config` is private
    |
    |
194 |         let config = Config { stage: 1, ..configure("build", &["A", "B"], &["A", "B"]) };
    |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ field `minimal_config` is private

error[E0451]: field `minimal_config` of struct `config::Config` is private
    |
    |
266 |         Config { stage: 2, ..super::configure("dist", host, target) }
    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ field `minimal_config` is private
For more information about this error, try `rustc --explain E0451`.
error: could not compile `bootstrap` due to 5 previous errors
Build completed unsuccessfully in 0:32:36
```

</details>

jyn514 commented on src/bootstrap/config.rs:
```diff
@@ -966,10 +969,10 @@ impl Config {
-            .map(PathBuf::from)
-            .unwrap_or_else(|| config.out.join(config.build.triple).join("stage0/bin/rustc"));
+        config.initial_rustc = build.rustc.map(PathBuf::from).unwrap_or_else(|| {
+            config.download_beta_toolchain();
```
ok, now that it's using `min_config` this no longer is downloaded in the shim

jyn514 commented:
This isn't ready to merge, but I think a review would be helpful to confirm that this is on the right track.

@rustbot ready

Mark-Simulacrum reviewed (commented):
I think in broad strokes this is looking OK to me. I do think we should avoid duplicating code if it can be avoided. That said I'm not sure I'm quite following what is driving duplication -- it might help to point me at it more explicitly.

I think we should start by just shipping this as opt-in, i.e., you have to do something to get it (perhaps even something awkward, like `alias x='rustup ...'` (with the command being the one from the hackmd), and get some testing in on a variety of setups before we flip the default.

bors commented:
:umbrella: The latest upstream changes (presumably #106409) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 commented:
@bors try

bors commented:
:hourglass: Trying commit 7cb16bd4998bbbb8143f99d0e4d582a69db3bc4e with merge 77c4fa259a0991578d13e8c98cc00f238873287d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7cb16bd4998bbbb8143f99d0e4d582a69db3bc4e","merge_sha":"77c4fa259a0991578d13e8c98cc00f238873287d"} -->

jyn514 commented:
@bors try

(there was a bug in the previous code that the latest commit fixes, the previous try build won't have `bootstrap-shim`)

bors commented:
:hourglass: Trying commit 443bbf085617bcbe313698940066e9d2d9622004 with merge 18b165dc185211a6c919bf0459866c88c35a6ca0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"443bbf085617bcbe313698940066e9d2d9622004","merge_sha":"18b165dc185211a6c919bf0459866c88c35a6ca0"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4032740675/jobs/6932812519) [(plain)](https://github.com/rust-lang/rust/commit/3ae67c50bbc91abb10ca9c87e3dad457436a233c/checks/10949129316/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
   Compiling thiserror-impl v1.0.38
   Compiling yaml-merge-keys v0.4.1
   Compiling expand-yaml-anchors v0.1.0 (/checkout/src/tools/expand-yaml-anchors)
    Finished release [optimized] target(s) in 4.53s
error: .github/workflows/ci.yml is not up to date; please run `x.py run src/tools/expand-yaml-anchors`.
caused by: src/ci/github-actions/ci.yml and .github/workflows/ci.yml are different
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #103019) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

onur-ozkan committed:
Unify `MinimalConfig::parse` and `Config::parse`

Signed-off-by: ozkanonur <work@onurozkan.dev>

ozkanonur commented:
Here is the commitment for `MinimalConfig::parse` and `Config::parse` unification https://github.com/ozkanonur/rust/commit/4fb924f1646175963e1d79b55a2f7d433b70b6cc

How do you want to handle this PR? @jyn514 You can cherry-pick that commit if you want(I guess this looks better way to deal with), or I can open another PR.

jyn514 commented:
@ozkanonur I'm happy for you to take over this PR :) I think I would prefer that to a new PR so the previous discussion is easy to find.

ozkanonur commented:
Could you please squash your commist in this PR? :slightly_smiling_face: I will reset my git history with your commit again (so we will have 2 commits(1 for your work, 1 for my) for the PR rather than 9 commits)

jyn514 committed:
Download beta compiler toolchain in bootstrap if it doesn't yet exist

This is needed for when the shell scripts bypass python altogether and run the downloaded bootstrap directly

jyn514 commented:
@ozkanonur done :) I left the first commit separate because it will also be needed if the python script starts downloading bootstrap, and I want to be able to extract it out into a different PR if necessary.

jyn514 committed:
Also re-download bootstrap if `build_helper` changes

jyn514 commented:
Ok, some more thoughts on the current state:

## Long-term plan

I am not sure if downloading bootstrap-shim from CI is the best approach. It seems silly to download two binaries one after the other, and not too hard to run `cargo run --bin bootstrap-shim` in the shell scripts. Compiling also avoids having to add bootstrap-shim to the rustup manifest, or needing to figure out exactly how old the shim can be before we force an update to a new version). Right now it doesn't actually save any time, because the shim depends on all of bootstrap's dependencies, but I think we can cut down the shim's dependencies to essentially 0 in a follow-up.

@Mark-Simulacrum do you have thoughts on the right approach here? I know this isn't the plan in the hackmd :( ... I can update it there if that's easier for you. Another thing we can do to avoid this PR bitrotting further is to merge this PR *without* changes to either the shell-scripts or CI, and have people use `cargo run --bin bootstrap-shim` manually to test out this workflow; that lets us defer the decision of exactly how to install it for later.

## Bugs

- Right now bootstrap-shim doesn't accept any arguments (oops):
```
; cargo r --bin bootstrap-shim --manifest-path src/bootstrap/Cargo.toml -- -v
    Finished dev [unoptimized] target(s) in 0.04s
     Running `/Users/jyn/.local/lib/cargo/target/debug/bootstrap-shim -v`
thread 'main' panicked at 'opts.parse(args) failed with Unrecognized option: 'v'', bin/bootstrap-shim.rs:12:19
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
This is non-trivial to fix with the current arg-parsing setup: either we'd have to share `flags.rs` between bootstrap and bootstrap-shim (which I'd rather not do to cut down on compile times), or we'd have to switch to a different arg-parsing library.

## Long-term bugs

These don't have to be fixed right away, but I think we should fix them before turning this on by default.

- The shim depends on all of bootstrap's dependencies. I think this is fixable, just really annoying because it means separating out "things bootstrap uses" from "things both bootstrap and bootstrap-shim use".
- We should add a `download-bootstrap # bool` option to config.toml to tell bootstrap-shim to fall back to using python instead.

jyn514 reviewed (commented):
@ozkanonur I see you made a partial start on unifying the parsing, but there's still some more logic I'd like to avoid duplicating - in particular, any option on `MinimalConfig` should only be set in min_config.rs, not config.rs. I see all the flags still being set in `config.rs`.

bors commented:
:umbrella: The latest upstream changes (presumably #108096) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 commented:
Closing in favor of https://github.com/rust-lang/rust/pull/107812
